Gif project [STEP1]

ehdwns0814 opened the pull request:
리뷰어: @forestjae
리뷰이: @ehdwns0814 

## 구현 사항
* MVVM 패턴 적용하여 ViewModel과 View, Controller로 구분 지어 구현
* apiKey xconfig파일로 관리하기

## 고민한 점
#### [3월 17일]
* Property Observer을 통해 데이터를 받으면 동적 데이터 처리와 데이터 처리 시 UI를 자연스럽게 처리 할 수 있어서 데이터 바인딩을 사용했습니다.
* HomeView와 검색어 입력 시 나오는 뷰가 collectionView 모양이 비슷하여 중복을 어떻게 줄일지 고민하였습니다.
#### [3월 31일]
* observable타입인 gif 저장 공간이 `storage`변수에 저장된 값에 접근하는 과정에서 고민하였습니다.
* url로 받아온 이미지가 움직이지 않아서 이미지를 움직이는 이미지로 변환 후에 다시 받아주었습니다.
* 움직이는 이미지를 반환 할때 비동기를 사용하지 않아 속도가 느려져서 비동기를 사용하였습니다.

## 궁금한 점
* 데이터 바인딩 방식에 Delegation 방식도 있는데 Property Observer과 무엇을 기준으로 데이터 바인딩 방식을 결정하는 지 알고 싶습니다.

ehdwns0814 committed:
feat: collectionView cell 생성

ehdwns0814 committed:
feat: collectionView 구현

ehdwns0814 committed:
feat: searchBar 검색 및 오토레이아웃 구현

ehdwns0814 committed:
feat: 이미지 요청 및 디코딩 구현

ehdwns0814 committed:
feat: 이미지 수신 및 Codingkey 구현

ehdwns0814 committed:
Config: 스토리보드 삭제

ehdwns0814 committed:
feat: 검색타입 enum 구현

ehdwns0814 committed:
feat: model 구현

ehdwns0814 committed:
feat: collectionViewCell 구현

ehdwns0814 committed:
feat: search viewModel 구현

ehdwns0814 committed:
feat: homeView 구현

ehdwns0814 committed:
feat: apiKey 숨기기 구현

ehdwns0814 committed:
feat: myViewController 생성

ehdwns0814 committed:
feat: observable 구현

데이터가 변동 시 View를 변동시키려고 구현

ehdwns0814 committed:
feat: network 요청 구현

ehdwns0814 committed:
feat: apiKey 불러오기 구현

ehdwns0814 committed:
feat: search화면 ViewModel 구현

ehdwns0814 committed:
feat: searchViewController 구현

ehdwns0814 committed:
feat: gifs 구조 구현

ehdwns0814 committed:
chore: 파일 삭제 및 이동

ehdwns0814 committed:
feat: tabBarViewController 구현

forestjae commented on ios-gif-search/Cells/GifCell.swift:
```diff
@@ -0,0 +1,39 @@
+class GifCell: UICollectionViewCell {
+    
+    static let identifier = "GifCell"
+    var gif: Gif?
```
private var

func configure(gif: GIF) {
   // 이미지 세팅 로직 
}

forestjae replied:
cell이 url을 알고있으니 이미지를 받아오는 역할까지 수행

forestjae commented on ios-gif-search/Cells/GifCell.swift:
```diff
@@ -0,0 +1,39 @@
+    static let identifier = "GifCell"
+    var gif: Gif?
+    
+    lazy var gifView: UIImageView = {
```
private

forestjae commented on ios-gif-search/Enums/ApiType.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+enum ApiType {
```
API

forestjae commented on ios-gif-search/Model/Gif.swift:
```diff
@@ -0,0 +1,30 @@
+        case gifSources = "images"
+    }
+    
+    func getGifURL() -> String {
+        return gifSources.original.url
+    }
```
```
var gifURL: String {
    return gifSources.original.url
}
```

forestjae commented on ios-gif-search/Search/SearchViewController.swift:
```diff
@@ -0,0 +1,125 @@
+class SearchViewController: UIViewController {
+    
+    let homeView = HomeView()
+    let network = GifNetwork(type: .search)
```
viewModel로 이동

forestjae commented on ios-gif-search/Search/SearchViewController.swift:
```diff
@@ -0,0 +1,125 @@
+        setupViews()
+        setupLayouts()
+        
+        viewModel.gifModels.bind { [weak self] _ in
+            DispatchQueue.main.async {
+                self?.collectionView.reloadData()
+            }
+        }
```
바인딩 로직은 다른 메서드로 분리해주세요

forestjae commented on ios-gif-search/Search/ViewModels/GifViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+import Foundation
+import UIKit
+
+struct GifListViewModel {
```
struct, class 어떤 차이가 있고, 왜 이걸 선택했는지 답변

forestjae commented on ios-gif-search/Search/ViewModels/GifViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+    var gifModels: Observable<[GifViewModel]> = Observable([])
+}
+
+extension GifListViewModel {
```
이름을 VC와 맞추기

forestjae replied:
extension을 꼭 나눠야하는지 생각해보기

forestjae commented on ios-gif-search/Search/ViewModels/GifViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+    
+    func numberOfItems() -> Int {
+        return gifModels.value!.count
```
! 사용 금지

forestjae commented on ios-gif-search/Search/ViewModels/GifViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+
+extension GifViewModel {
+    func fetchImage(completion: @escaping (UIImage?) -> Void) {
+        guard let url = URL(string: url) else {
+            completion(nil)
+            return
+        }
+        
+        URLSession.shared.dataTask(with: url) { data, response, error in
+            guard let data = data, error == nil, let image = UIImage(data: data) else {
+                completion(nil)
+                return
+            }
+            completion(image)
+        }.resume()
```
공통 유틸리티 객체 분리

forestjae commented on ios-gif-search/Search/ViewModels/GifViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+//
+//
+import Foundation
+import UIKit
```
UIKit View에서만 import 하도록

forestjae commented on ios-gif-search/Search/SearchViewController.swift:
```diff
@@ -0,0 +1,125 @@
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: GifCell.identifier, for: indexPath) as! GifCell
+//        cell.gif = viewModel.gifAtIndex(indexPath.row)
```
cell.configure(gif: viewModel. gifModels.value[indexPath.row])

forestjae commented on ios-gif-search/Services/GifNetwork.swift:
```diff
@@ -0,0 +1,60 @@
+
+import Foundation
+
+class GifNetwork {
```
GifService

forestjae replied:
final

ehdwns0814 committed:
refactor: 변수명 변경

Api -> API

ehdwns0814 committed:
feat: ViewModelProtocol 구현

ehdwns0814 committed:
refactor: 네이밍 변경

GifNetwork -> GifService 변경

ehdwns0814 committed:
feat: 변수명 변경 및 메서드 파라미터 변경

- GifViewModel -> SearchViewModel 네이밍 변경
- fetchData(): enum 타입 파라미터로 변경
- setError(): 에러 타입 설정
- storage 프로퍼티 추가 데이터 저장목록 생성

ehdwns0814 committed:
feat: enum타입 구현

- GifMenu: 데이터 타입 종류
- GifType: 검색 종류

ehdwns0814 committed:
feat: collectionCell 구현

- configure(): cell imageView에 이미지 설정

ehdwns0814 committed:
refactor: Observable 타입 수정

ehdwns0814 committed:
refactor: 연산 프로퍼티로 변경

ehdwns0814 committed:
feat: Error 타입 구현

ehdwns0814 committed:
feat: 데이터 바인딩 구현

- network: viewModel로 이동
- setupData(): 데이터를 저장소에 저장
- setupBinding(): 저장된 데이터를 collectionView에 바인딩

ehdwns0814 committed:
chore: 중복파일 삭제

ehdwns0814 committed:
chore: 파일 이동 및 삭제 저장

ehdwns0814 committed:
feat: Error 구현

- error 케이스 추가

ehdwns0814 committed:
feat: collectionViewCell URL이미지 gif로 움직이게 구현

- UIImage: extension으로 이미지를 움직이는 이미지로 반환
- configure(): url으로 이미지를 요청하여 받아온후 이미지를 움직이는 이미지로 변환 후 cell의 이미지에 대입

ehdwns0814 committed:
feat: 받아온 데이터 storage에 저장

- fetchData(): 원하는 타입의 url로 검색 요청 후 받은 url들을 storage에 저장

ehdwns0814 committed:
feat: 데이터에 바인딩하여 데이터 변경 시 View 변경

- storage의 데이터가 변경 될 시에 collectionView의 이미지를 변경

ehdwns0814 committed:
chore: 변수명 변경, 불필요한 코드 제거, 파일 위치 변경
